### ISSUE TITLE
Serialization of email with '@' in the name in NameEmail

### ISSUE THREAD

**Issue by littledumb** [pydantic/pydantic#8811]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Serializing a NameEmail object with '@' in the name should wrap the name in double or single quotes, otherwise deserializing the string back to NameEmail will cause an error.

### Example Code

```Python
import json
from pydantic import BaseModel, NameEmail

class Email(BaseModel):
  From: NameEmail

email = Email(From='"name@mailbox.com" <name@mailbox.com>')
print(email)
# output
# From=NameEmail(name='name@mailbox.com', email='name@mailbox.com')

obj = json.loads(email.model_dump_json())
print(obj)
# output
# {'From': 'name@mailbox.com <name@mailbox.com>'}

email = Email(From=obj['From'])
# Traceback (most recent call last):
#  ...
#    email = Email(From=obj['From'])
#            ^^^^^^^^^^^^^^^^^^^^^^^
#  File "/<some_path>/lib/python3.11/site-packages/pydantic/main.py", line 171, in __init__
#    self.__pydantic_validator__.validate_python(data, self_instance=self) pydantic_core._pydantic_core.ValidationError: 1 validation error for Email From
#  value is not a valid email address: The email address is not valid. It must have exactly one @-sign. [type=value_error, input_value='name@mailbox.com <name@mailbox.com>', input_type=str]
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.6.1
        pydantic-core version: 2.16.2
          pydantic-core build: profile=release pgo=false
                 install path: /<some_path>/.local/share/virtualenvs/<some_project_name>/lib/python3.11/site-packages/pydantic
               python version: 3.11.7 (main, Jan 16 2024, 15:02:38) [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-12.5.1-x86_64-i386-64bit
             related packages: typing_extensions-4.9.0 pydantic-settings-2.1.0 email-validator-2.1.0.post1
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
@littledumb,

Related: https://github.com/pydantic/pydantic/issues/3173
PR welcome with a fix, but we'll want to verify that having a name with the `@` symbol is valid according to https://datatracker.ietf.org/doc/html/rfc5322#section-3.4

**Comment by littledumb** [comment]
@sydney-runkle 

AWS WorkMail adds the mailbox address (in double quotes) as a display name by default.  The address string passes our email routing process. However, our downstream application reports error when loading the serialized address. We have to add a workaround on serialization of the address.

**Comment by NeevCohen** [comment]
Hey @sydney-runkle I can try and take a shot at this, if no one's working on it yet of course

**Comment by NeevCohen** [comment]
@sydney-runkle I couldn't quite figure out if this format of addressing is valid according to the RFC, perhaps I'm just missing something. 

Anyway, a possible solution for this would could be in the `__str__` method of `NameEmail`

```python
def __str__(self) -> str:
    if pretty_email_regex.fullmatch(self.name):
        return f'"{self.name}" <{self.email}>'
    
    return f'{self.name} <{self.email}>'
```

**Comment by NeevCohen** [comment]
After messaging the author of the RFC it seems that syntactically, a mailbox name wrapped in double quotes is perfectly valid as a display name, Though not recommended since many mail clients only show the display name to the user. So, someone could send an email from "support@gmail.com" <badguy@spammer.org>.

That being said, looks like this is a valid display name. @sydney-runkle Do you think the change I suggested would be a viable fix? It doesn't cover #3173 though

**Comment by sydney-runkle** [comment]
Closed by the above ^^
